Ensure the 'inputs' field in /chat-messages takes effect every time

Wu-Jiayang opened the pull request:
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Close #7952
Close https://github.com/langgenius/dify/issues/7846
Close https://github.com/langgenius/dify/issues/8063
Close https://github.com/langgenius/dify/issues/8839
Close https://github.com/langgenius/dify/issues/8546
Close https://github.com/langgenius/dify/issues/10522
Close https://github.com/langgenius/dify/issues/10055
Close https://github.com/langgenius/dify/pull/10643
Close https://github.com/langgenius/dify/issues/13363

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

Commit:
Ensure the 'inputs' field in /chat-messages takes effect every time

Commit:
Merge remote-tracking branch 'origin'

Commit:
Ensure the 'inputs' field in /chat-messages takes effect every time

Commit:
Merge remote-tracking branch 'origin'

cubxxw commented:
This PR does not resolve, using the conversation id to link to previous sessions but the input works

Wu-Jiayang commented:
> This PR does not resolve, using the conversation id to link to previous sessions but the input works

I don't understand "using the conversation id to link to previous sessions but the input works". As for "This PR does not resolve", I have tried and I think it works.

Wu-Jiayang commented:
> This PR does not resolve, using the conversation id to link to previous sessions but the input works

This PR is to make sure the 'inputs' field in the API "/chat-messages" takes effect every time, is that what you mean "using the conversation id to link to previous sessions but the input works"?

cubxxw commented:
Yes, you mean that the input field of the conversation can be updated under the same conversation, because the input is only effective when the conversation link is first established

Wu-Jiayang commented:
> Yes, you mean that the input field of the conversation can be updated under the same conversation, because the input is only effective when the conversation link is first established

@cubxxw 

I noticed that both table 'messages' and table 'conversations' have an 'inputs' field. I think the 'inputs' field should be aligned with table 'messages'. The 'inputs' field in table 'conversations' may not be very useful.

In this PR, the 'inputs' field in table 'conversations' means the 'inputs' in the first message, and the 'inputs' field in table 'messages' means the true inputs in this message.

Wu-Jiayang commented:
> This PR does not resolve, using the conversation id to link to previous sessions but the input works

What's the mean of 'This PR does not resolve'? Is that any problems?

cubxxw commented:
https://github.com/langgenius/dify/issues/8063

This is the problem I encountered

cubxxw commented:
I understand that the conversation table input is valid for the first time. The message table is used for every subsequent interaction, right?

Wu-Jiayang commented:
> #8063
> 
> This is the problem I encountered,
> 
> maybe I understand wrong, this PR can't solve the problem of the video in issue?

This PR can solve the problem of the video in issue. @cubxxw 

https://github.com/user-attachments/assets/78f2b50a-1688-436f-b8dd-eb6153a5bb51

Wu-Jiayang commented:
> I understand that the conversation table input is valid for the first time. The message table is used for every subsequent interaction, right?

right

Hisir0909 reviewed (approved):
Excellent changes, making the workflow much more flexible.

cubxxw commented:
I think it's great, too

Hope to support this feature

liuzixiao666 commented:
Has this feature not been merged？

Wu-Jiayang commented:
> Has this feature not been merged？

I don't know why they don't merge it, maybe we need more people to push them.

Wu-Jiayang committed:
Merge branch 'main' into main

liuzixiao666 commented:
> Author

I highly appreciate your functionality and have submitted a new issue. If they do not merge, I will download your branch
[https://github.com/langgenius/dify/issues/10522](url)

crazywoola commented:
I have already talk to our team about this pr. :)
We will evaluate this later.

crazywoola commented:
I have opened a discussion here, please forward this to other community groups, we would like to gather some information regarding this feature. 

https://github.com/langgenius/dify/discussions/10652

cubxxw commented:
Hope to merge

liuzixiao666 commented:
I tried to merge your code in my project, but it didn't take effect. I modified the three lines of code in those three files and changed them to inputs=self_ Prepare_user_inputs (user_inputs=inputs, app_ config=app_ config), right?

Wu-Jiayang commented:
> I tried to merge your code in my project, but it didn't take effect. I modified the three lines of code in those three files and changed them to inputs=self_ Prepare_user_inputs (user_inputs=inputs, app_ config=app_ config), right?

right

liuzixiao666 commented:
> > I tried to merge your code in my project, but it didn't take effect. I modified the three lines of code in those three files and changed them to inputs=self_ Prepare_user_inputs (user_inputs=inputs, app_ config=app_ config), right?
> 
> right

> 

Uploading 无标题视频——使用Clipchamp制作.mp4…

liuzixiao666 commented:
https://github.com/user-attachments/assets/12839e0c-9c66-443c-8737-fba1158b631e

Wu-Jiayang committed:
Merge branch 'langgenius:main' into main

liuzixiao666 commented:
> > It seems to have failed? My DIY version is 0.11.0似乎失败了？我的DIY版本是0.11.0
> 
> @liuzixiao666 I merged the 'langgenius:main' branch into my branch and tested it on v0.11.1. It works well.我将"langgenius：main“分支合并到我的分支中，并在v0.11.1上进行了测试。很好用。
> 
>  2024-11-15.17-36-33.mp4

i will have a try again

liuzixiao666 commented:
> > It seems to have failed? My DIY version is 0.11.0
> 
> @liuzixiao666 I merged the 'langgenius:main' branch into my branch and tested it on v0.11.1. It works well.
> 
>  2024-11-15.17-36-33.mp4

Is mine still not effective? Is it because I started Docker? Are you starting from local source code

Wu-Jiayang commented:
> > > It seems to have failed? My DIY version is 0.11.0
> > 
> > 
> > @liuzixiao666 I merged the 'langgenius:main' branch into my branch and tested it on v0.11.1. It works well.
> > 2024-11-15.17-36-33.mp4
> 
> Mine still hasn't taken effect. Is it because I started Docker? Are you starting from local source code

I built Docker from local source code and run `docker run -d -e MODE=api --restart=unless-stopped --name "dify-api" --network host --env-file .env dify-api`

liuzixiao666 commented:
> > > > It seems to have failed? My DIY version is 0.11.0
> > > 
> > > 
> > > @liuzixiao666 I merged the 'langgenius:main' branch into my branch and tested it on v0.11.1. It works well.
> > > 2024-11-15.17-36-33.mp4
> > 
> > 
> > Mine still hasn't taken effect. Is it because I started Docker? Are you starting from local source code
> 
> I built Docker from local source code and run `docker run -d -e MODE=api --restart=unless-stopped --name "dify-api" --network host --env-file .env dify-api`

I can't quite understand your code, it seems like you have rebuilt a new Docker container. Does that mean that stopping the container and using 'docker compose up - d' will not work when I run it in the Docker directory? Do you know how to make the modified code take effect

Wu-Jiayang commented:
@liuzixiao666 
I have read the default `docker-compose.yaml` and It seems not from local source code:
![image](https://github.com/user-attachments/assets/dd8fed92-6311-44be-87d0-398e8184c02e)

If you want to use docker, you need to make sure the docker you use is built locally, or waiting for merging this PR. ^_^

TheFu527 commented:
What a great PR! This will solve a lot of problems! Looking forward to merging

liuzixiao666 commented:
> What a great PR! This will solve a lot of problems! Looking forward to merging

Remember to give a like here 
#10652

HaiShin commented:
> > > > It seems to have failed? My DIY version is 0.11.0
> > > 
> > > 
> > > @liuzixiao666 I merged the 'langgenius:main' branch into my branch and tested it on v0.11.1. It works well.
> > > 2024-11-15.17-36-33.mp4
> > 
> > 
> > Mine still hasn't taken effect. Is it because I started Docker? Are you starting from local source code
> 
> I built Docker from local source code and run `docker run -d -e MODE=api --restart=unless-stopped --name "dify-api" --network host --env-file .env dify-api`

I think your three lines of code will not work, because each conversation does not pass the inputs from the front end. My version is 0.11.2.

Wu-Jiayang commented:
@HaiShin 
I tested my code on v0.11.1. It works well.

Wu-Jiayang committed:
Merge branch 'main' into main

liuzixiao666 commented:
> @HaiShin I tested my code on v0.11.1. It works well. I don't know why you said 'each conversation does not pass the inputs from the front end'?

do u have wechatid？I am unable to wait for the official merge PR and would like to know how to build Docker locally

liuzixiao666 commented:
> @liuzixiao666 I have read the default `docker-compose.yaml` and It seems not from local source code: ![image](https://private-user-images.githubusercontent.com/62842862/386823319-dd8fed92-6311-44be-87d0-398e8184c02e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3NzI5MjMsIm5iZiI6MTczMjc3MjYyMywicGF0aCI6Ii82Mjg0Mjg2Mi8zODY4MjMzMTktZGQ4ZmVkOTItNjMxMS00NGJlLTg3ZDAtMzk4ZTgxODRjMDJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI4VDA1NDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0NTJkMzY4ZWI1NzY4MWU0OThhNDg0YjliNjBmOWY2NGNiYTM1YWJlZmFmYjlhNGM4ZjM2M2Q4ODNhNzcwN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.O73D9x6RlHwPyKbN0nsqqRbTvmEkWR3DeWBCYMTO-xo)
> 
> If you want to use docker, you need to make sure the docker you use is built locally, or waiting for merging this PR. ^_^

do u have wechatid？I am unable to wait for the official merge PR and would like to know how to build Docker locally

liuzixiao666 commented:
@Wu-Jiayang do u have wechatid（vx）？I am unable to wait for the official merge PR and would like to know how to build Docker locally

Wu-Jiayang commented:
> @Wu-Jiayang do u have wechatid（vx）？I am unable to wait for the official merge PR and would like to know how to build Docker locally

https://docs.dify.ai/zh-hans/getting-started/install-self-hosted/local-source-code

```sh
cd web
docker build . -t dify-web
docker run -d -p 3000:3000 --restart=unless-stopped --name "dify-web" -e CONSOLE_API_URL=http://108.61.191.62:5001 -e APP_API_URL=http://108.61.191.62:5001 dify-web

cd docker
cp middleware.env.example middleware.env
docker compose -f docker-compose.middleware.yaml up -d

cd api
docker build . -t dify-api
cp .env.example .env  # 生成.env
# 生成随机密钥，并替换 .env 中 SECRET_KEY 的值
awk -v key="$(openssl rand -base64 42)" '/^SECRET_KEY=/ {sub(/=.*/, "=" key)} 1' .env > temp_env && mv temp_env .env
# 别忘了配置.env
docker run -d -e MODE=api --restart=unless-stopped --name "dify-api" --network host --env-file .env dify-api
docker run -d -e MODE=worker -e CELERY_WORKER_AMOUNT=2 --restart=unless-stopped --name "dify-worker" --network host --env-file .env dify-api
```

liuzixiao666 commented:
> > @Wu-Jiayang do u have wechatid（vx）？I am unable to wait for the official merge PR and would like to know how to build Docker locally
> 
> https://docs.dify.ai/zh-hans/getting-started/install-self-hosted/local-source-code
> 
> ```shell
> cd web
> docker build . -t dify-web
> docker run -d -p 3000:3000 --restart=unless-stopped --name "dify-web" -e CONSOLE_API_URL=http://108.61.191.62:5001 -e APP_API_URL=http://108.61.191.62:5001 dify-web
> 
> cd docker
> cp middleware.env.example middleware.env
> docker compose -f docker-compose.middleware.yaml up -d
> 
> cd api
> docker build . -t dify-api
> cp .env.example .env  # 生成.env
> # 生成随机密钥，并替换 .env 中 SECRET_KEY 的值
> awk -v key="$(openssl rand -base64 42)" '/^SECRET_KEY=/ {sub(/=.*/, "=" key)} 1' .env > temp_env && mv temp_env .env
> # 别忘了配置.env
> docker run -d -e MODE=api --restart=unless-stopped --name "dify-api" --network host --env-file .env dify-api
> docker run -d -e MODE=worker -e CELERY_WORKER_AMOUNT=2 --restart=unless-stopped --name "dify-worker" --network host --env-file .env dify-api
> ```

I have tried before, but encountered too many problems in the middle and gave up. Alright, I understand the way you built it. Thank you

liuzixiao666 commented:
When can we merge PR                    o(╥﹏╥)o

Yoha485 commented:
Guys, let's merge the PR

Wu-Jiayang commented:
> I have already talked to our team about this pr. :) We will evaluate this later.

@crazywoola

shenyi0828 commented:
Guys, it's 0.14.1 ! Still not merged? Why?  "inputs" is madatory each call , but not updated when conversation_id is provided??? and still recorded to DB every time in table "messages" . That's a BUG. please fix it

dc-vin commented:
Has this bug not been fixed and merged after so many versions for any particular reason? The issue I submitted was closed, but the bug still persists.

Wu-Jiayang commented:
I'm just this PR's Author, not the project's member, please ask this guy @crazywoola why. ^_^

Wu-Jiayang commented:
> I have already talked to our team about this pr. :) We will evaluate this later.

When to review my code? It's only three lines... @crazywoola

cubxxw commented:
@crazywoola @takatost  Whether there is a specific planned merger time

liuzixiao666 commented:
From version 0.10.0 to 0.15.0, it still hasn't been resolved

Wu-Jiayang commented:
@crazywoola @takatost

juniorsereno commented:
We need this PR, hello admin

cubxxw commented:
Version 1.0 ？？？

Yuyang105 commented:
Merge, please! This is a total buzzkill, and now the workflow won't flow anymore. Save me from my DDL by the end of lunar new year :[

liuzixiao666 commented:
> Merge, please! This is a total buzzkill, and now the workflow won't flow anymore. Save me from my DDL by the end of lunar new year :[

No one is paying attention to us, and this PR cannot be merged for a long time. My product has also been stuck in this area and I have to seek alternative products to replace it

Yuyang105 commented:
> > Merge, please! This is a total buzzkill, and now the workflow won't flow anymore. Save me from my DDL by the end of lunar new year :[
> 
> No one is paying attention to us, and this PR cannot be merged for a long time. My product has also been stuck in this area and I have to seek alternative products to replace it

I decided to bypass this issue by sending http request to get 'inputs' from external storage, and then adding a code execution node to parse the context variables, achieving what was supposed to be dynamic. 🙂‍↕️

liuzixiao666 commented:
> > > Merge, please! This is a total buzzkill, and now the workflow won't flow anymore. Save me from my DDL by the end of lunar new year :[
> > 
> > 
> > No one is paying attention to us, and this PR cannot be merged for a long time. My product has also been stuck in this area and I have to seek alternative products to replace it
> 
> I decided to bypass this issue by sending http request to get 'inputs' from external storage, and then adding a code execution node to parse the context variables, achieving what was supposed to be dynamic. 🫨

We have also discussed this solution, which is only a short-term alternative and requires identity recognition for some unique values of the user, as well as providing accessible interfaces and more complex workflow modifications

Wu-Jiayang commented:
@crazywoola @takatost @laipz8200 @iamjoel

abnormalboy commented:
i need this pr

cubxxw commented:
Happy New Year, but what?

Wu-Jiayang commented:
Happy New Year, everybody! @crazywoola @takatost @laipz8200 @iamjoel @JohnJyong @zxhlyh

crazywoola commented:
I'll revisit this with the team later, but we probably won't be releasing a minor update anytime soon, and we're working on a 1.0 update.

Wu-Jiayang commented:
We've already been waiting for five months, and we don't mind waiting a little longer, as long as it won't be another five months.
^_^

liuzixiao666 commented:
> We've already been waiting for five months, and we don't mind waiting a little longer, as long as it won't be another five months. ^_^

hhhhhhhhhhhhhhhhhhh，maybe 5 years

laipz8200 reviewed (approved):
LGTM

laipz8200 committed:
Merge branch 'main' into Wu-Jiayang/main

deific commented:
LGTM

Wu-Jiayang commented:
when to merge?

cubxxw commented:
If we don't merge, I wonder if there's any way to add the DeepSeek R1 model. The current code is not easy to update.？

laipz8200 commented:
> If we don't merge, I wonder if there's any way to add the DeepSeek R1 model. The current code is not easy to update.？

This PR is not about the deepseek models, we provider many integrations of deepseek in the 0.15.3 version.

Check the [release note](https://github.com/langgenius/dify/releases/tag/0.15.3).

dc-vin commented:
> > If we don't merge, I wonder if there's any way to add the DeepSeek R1 model. The current code is not easy to update.？
> 
> This PR is not about the deepseek models, we provider many integrations of deepseek in the 0.15.3 version.
> 
> Check the [release note](https://github.com/langgenius/dify/releases/tag/0.15.3).

How to not merge this PR, how can I switch the DeepSeek model in a session instead of using different applications to implement the switch, the chat window on the DeepSeek official website can switch between DeepSeek-R1 and DeepSeek-V3.

dc-vin commented:
> > If we don't merge, I wonder if there's any way to add the DeepSeek R1 model. The current code is not easy to update.？
> 
> This PR is not about the deepseek models, we provider many integrations of deepseek in the 0.15.3 version.
> 
> Check the [release note](https://github.com/langgenius/dify/releases/tag/0.15.3).

Additionally, how can I switch between DeepSeek-R1 and DeepSeek-V3 models within a session, rather than using different applications to implement the switch? I noticed that the chat window on the DeepSeek official website allows switching between DeepSeek-R1 and DeepSeek-V3

MJSanfelippo commented:
Please merge this, we need this feature asap

Halfknow commented:
Please merge this, we need this feature asap

liuzixiao666 commented:
Please merge this, we need this feature asap

Wu-Jiayang commented:
@crazywoola @takatost @laipz8200 @iamjoel @JohnJyong @zxhlyh

MJSanfelippo commented:
> @crazywoola @takatost @laipz8200 @iamjoel @JohnJyong @zxhlyh

Owners of Dify, please merge and release this. It's holding us back from using Dify.

In the meantime, anyone know any other tools like Dify?

Wu-Jiayang commented:
maybe ragflow or fastgpt

cubxxw commented:
lgtm

bowenliang123 commented:
Great !

MJSanfelippo commented:
@crazywoola  when will this be released?

chiehw commented:
This PR breaks the original inputs design. https://github.com/langgenius/dify-docs-mintlify/blob/78095626e30661f57a539edd31729aa3be096770/zh-hans/guides/application-publishing/developing-with-apis.mdx#conversation_id-%E7%9A%84%E6%B3%A8%E6%84%8F%E4%BA%8B%E9%A1%B9

Wu-Jiayang commented:
![image](https://github.com/user-attachments/assets/840ab540-5f0a-4a0b-bd31-6385605d1005)

They have already talked to their team about this pr and discussed for a long time.
